[Enhancement] schema change optimization for adding generated column (#24728)

srlch opened the pull request:
Fixes #24728
Use the delta column group to optimize adding a generated column

We just test the feature on a table containing some huge JSON columns with 6 million rows of data
The result indicates that if the table has 5/10 JSON columns, and the generated column's expression
is just associated with only one JSON column, the schema change speed can be optimized by 9 to 18
times respectively.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

andyziye commented:
@Mergifyio rebase

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

chaoyli commented on be/src/runtime/snapshot_loader.cpp:
```diff
@@ -655,7 +655,13 @@ Status SnapshotLoader::move(const std::string& snapshot_path, const TabletShared
         // link files one by one
         // files in snapshot dir will be moved in snapshot clean process
         std::vector<std::string> linked_files;
+        std::string dcg_file;
```
It's better to describe the background, solution and effects in the commit message.

chaoyli commented on be/src/runtime/snapshot_loader.cpp:
```diff
@@ -655,7 +655,13 @@ Status SnapshotLoader::move(const std::string& snapshot_path, const TabletShared
+        std::string dcg_file;
         for (auto& file : snapshot_files) {
+            // A tablet can have at most one dcgs snapshot file
+            if (_end_with(file, ".dcgs_snapshot")) {
```
1. If the schema change failed when to clear the `.dcgs_snapshot`
2. If the schema change failed and the BE restarted, when to clear the `.dcgs_snapshot`?

srlch replied:
Here is the restore related code. .dcgs_snapshot is the file contain the metadata of the backup table's dcg information. The . dcgs_snapshot file is always in snapshot path, which will be remove later in the restore procedure.

chaoyli commented on be/src/runtime/snapshot_loader.cpp:
```diff
@@ -673,6 +679,44 @@ Status SnapshotLoader::move(const std::string& snapshot_path, const TabletShared
+                }
+                st = data_dir->get_meta()->write_batch(&wb);
+                if (!st.ok()) {
+                    return Status::InternalError("save dcgs meta failed");
```
Print the tablet_id in the error log to better trace.

srlch replied:
done

chaoyli commented on be/src/storage/schema_change.cpp:
```diff
@@ -888,10 +1017,83 @@ Status SchemaChangeHandler::_convert_historical_rowsets(SchemaChangeParams& sc_p
+            RETURN_IF_ERROR(TabletMetaManager::get_delta_column_group(new_tablet->data_dir()->get_meta(), tablet_id,
+                                                                      rowsetid, j, INT64_MAX, &historical_dcgs[j]));
+        }
+        if (!sc_params.sc_sorting && !sc_params.sc_directly && chunk_changer->get_mc_exprs()->size() != 0) {
```
`gc` not `mc`. `mc` is short for `materialized column`

srlch replied:
I will fix the naming problem in another pr for both be and fe

chaoyli commented on be/src/storage/schema_change.cpp:
```diff
@@ -888,10 +1017,83 @@ Status SchemaChangeHandler::_convert_historical_rowsets(SchemaChangeParams& sc_p
+            if (dcgs.size() != 0) {
+                if (dcgs.size() != sc_params.rowsets_to_change[i]->num_segments()) {
+                    std::stringstream ss;
+                    ss << "The size of dcgs and segment file in src rowset is different.";
```
Add `tablet_id` in the log

srlch replied:
done

chaoyli commented on be/src/storage/schema_change.cpp:
```diff
@@ -888,10 +1017,83 @@ Status SchemaChangeHandler::_convert_historical_rowsets(SchemaChangeParams& sc_p
+    if (!status.ok()) {
+        LOG(WARNING) << "Fail to delete old dcg and write new dcg" << sc_params.new_tablet->tablet_id() << ": "
+                     << status;
+        return Status::InternalError("Fail to delete old dcg and write new dcg");
```
Add tablet_id in the log

srlch replied:
done

chaoyli commented on be/src/storage/schema_change.cpp:
```diff
@@ -269,6 +270,133 @@ Status LinkedSchemaChange::process(TabletReader* reader, RowsetWriter* new_rowse
 }
 
+Status LinkedSchemaChange::convert_from_base_rowset_by_cols(const Tablet* new_tablet, const Tablet* base_tablet,
+                                                            const RowsetSharedPtr src_rowset, RowsetId rid,
```
const TabletPtr& instead of pointer. pointer means to change the content.

srlch replied:
I think we should keep const Tablet* because the function call convert_from_base_rowset_by_cols is just raw point Tablet* but not share pointer

chaoyli replied:
The shared_point is the same usage with the naked pointer.

srlch commented on be/src/storage/rowset/rowset.cpp:
```diff
@@ -360,18 +370,32 @@ Status Rowset::_link_delta_column_group_files(const std::string& dir, int64_t ve
+                    const std::string& dst_link_path = dst_link_paths[j];
+
+                    if (link(src_file_path.c_str(), dst_link_path.c_str()) != 0) {
+                        PLOG(WARNING) << "Fail to link " << src_file_path << " to " << dst_link_path;
```
fixed

srlch commented on be/src/storage/storage_engine.h:
```diff
@@ -78,6 +78,34 @@ class CompactionManager;
+            return false;
+        } else if (rowsetid < rhs.rowsetid) {
+            return true;
+        } else if (rowsetid != rhs.rowsetid) {
```
RowsetId does not define > operator

srlch commented on be/src/storage/schema_change.cpp:
```diff
@@ -888,10 +1017,83 @@ Status SchemaChangeHandler::_convert_historical_rowsets(SchemaChangeParams& sc_p
+                                                     ? historical_dcgs[j].front()->relative_column_files().size()
+                                                     : 0);
+            }
+            status = LinkedSchemaChange::convert_from_base_rowset_by_cols(
```
done

srlch commented on be/src/storage/schema_change.cpp:
```diff
@@ -269,6 +270,133 @@ Status LinkedSchemaChange::process(TabletReader* reader, RowsetWriter* new_rowse
+                    return Status::InternalError(ss.str());
+                }
+            }
+            status = chunk_changer->append_materialized_columns(read_chunk, new_chunk, all_ref_columns_ids,
```
fixed in later pr

srlch commented on be/src/storage/schema_change.cpp:
```diff
@@ -269,6 +270,133 @@ Status LinkedSchemaChange::process(TabletReader* reader, RowsetWriter* new_rowse
     return Status::OK();
 }
 
+Status LinkedSchemaChange::convert_from_base_rowset_by_cols(const Tablet* new_tablet, const Tablet* base_tablet,
```
done

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

luohaha commented on be/src/storage/rowset/segment.cpp:
```diff
@@ -218,15 +218,24 @@ Status Segment::_open(size_t* footer_length_hint, const FooterPointerPB* partial
+        st = read_options.dcg_loader->load(tsid, read_options.version, &dcgs);
+    } else {
+        int64_t tablet_id = read_options.tablet_id;
+        RowsetId rowsetid = read_options.rowsetid;
```
`rowset_id` is too close to `rowsetid`. Change this name?

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

luohaha commented on be/src/storage/rowset/rowset.cpp:
```diff
@@ -144,8 +144,8 @@ std::string Rowset::segment_upt_file_path(const std::string& dir, const RowsetId
-                                            int64_t version) {
-    return strings::Substitute("$0/$1_$2_$3.cols", dir, rowset_id.to_string(), segment_id, version);
+                                            int64_t version, int idx) {
+    return strings::Substitute("$0/$1_$2_$3_$4.cols", dir, rowset_id.to_string(), segment_id, version, idx);
```
Will there is a compatible problem here ? if old `.cols` is `$0/$1_$2_$3.cols` ?

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

srlch committed:
[Enhancement] schema change optimization for adding generated column (#24728)

Use delta column group to optimize adding generated column

We just test the feature on a table containing some huge JSON column with 6 million rows of data
The result indicate that if the table has 5/10 JSON columns, and the generated column's expression
is just aossciated with only one JSON column, the schema change speed can be optimized by 9 to 18
times respectively.

Signed-off-by: srlch <linzichao@starrocks.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)
